File: 3rd Year/Software Engineering/LowestCommonAncestor/LowestCommonAncestor.py
class Node:
    def __init__(self, key):
       self.left = None
       self.right = None
       self.key = key

def find_path(root, path, k):
    if root is None:
        return False
    
    path.append(root.key)
    if(root.key == k):
        return True
    if((root.left != None and find_path(root.left, path, k)) or (root.right != None and find_path(root.right, path, k))):
       return True
    path.pop()
    return False
def LCA(root, x, y):
    p1 = []
    p2 = []
    if(not find_path(root, p1, x) or not find_path(root, p2, y)):
        return -1
    k = 0
    while(k < len(p1) and k < len(p2)):
        if(p1[k] != p2[k]):
            break
        k = k + 1
    return p1[k-1]

File: 3rd Year/Software Engineering/LowestCommonAncestor/LowestCommonAncestor.js
class Node {
  constructor(key) {
    this.left = null;
    this.right = null;
    this.key = key;
  }
}

const LCA = (root, p, q) => {
  const LCA_inner = (root, p, q) => {
    if (root == null) return root;
    if (root.key == p || root.key == q) return root;
    var left = LCA_inner(root.left, p, q);
    var right = LCA_inner(root.right, p, q);
    if (left != null && right != null) return root;
    if (left == null && right == null) return null;
    return left != null ? left : right;
  };
  return LCA_inner(root, p, q).key;
};

let root = new Node(1);
root.left = new Node(2);
root.right = new Node(3);
root.left.left = new Node(4);
root.left.right = new Node(5);
root.right.left = new Node(6);
root.right.right = new Node(7);
a1 = LCA(root, 4, 5);
a2 = LCA(root, 4, 6);
a3 = LCA(root, 3, 4);
a4 = LCA(root, 2, 4);
console.log(a1, a2, a3, a4);
